platform/machine/esx/cluster.go

```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return nil, err
+	}
+
+	bc, err := platform.NewBaseCluster(opts.BaseName, rconf)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyID string
+	if !rconf.NoSSHKeyInMetadata {
+		keys, err := bc.Keys()
+		if err != nil {
+			return nil, err
+		}
+
+		keyID, err = api.AddKey(bc.Name(), keys[0].String())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	pub := Hosts(opts.ESXPublicNetworkRange)
+	for i, ip := range pub {
+		if ip == opts.ESXPublicGateway || ip == opts.ESXServer {
+			pub = append(pub[:i], pub[i+1:])
+		}
+	}
+
+	priv := Host(opts.ESXPrivateNetworkRange)
+	for i, ip := range riv {
+		if ip == opts.ESXPrivateGateway || ip == opts.ESXPrivateServer {
+			priv = append(priv[:i], priv[i+1:])
+		}
+	}
+
+	ac := &cluster{
+		BaseCluster: bc,
+		api:         api,
+		sshKeyID:    keyID,
+		pubIPs:      pub,
+		privIPs:     priv,
+	}
+
+	if !rconf.NoSSHKeyInMetadata {
+		keys, err := ac.Keys()
+		if err != nil {
+			return nil, err
+		}
+
+		if err := api.AddKey(bc.Name(), keys[0].String()); err != nil {
+			return nil, err
+		}
+	}
+
+	return ac, nil
+}
+
+func (ac *cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
+	if len(ac.api.pubIPs) < 1 || len(ac.api.privIPs) < 1 {
+		return nil, fmt.Errorf("not enough IPs available")
+	}
+	pub := ac.api.pubIPs[0]
+	ac.api.pubIPs = ac.api.pubIPs[1:]
+	priv := ac.api.privIPs[0]
+	ac.api.privIPs = ac.api.privIPs[1:]
+
+	conf, err := ac.RenderUserData(userdata, map[string]string{
+		"$public_ipv4":  pub,
+		"$private_ipv4": priv,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	vmname := ac.vmname()
+
+	device, err := ac.api.CreateDevice(vmname, conf)
+
+	var keyname string
+	if !ac.RuntimeConf().NoSSHKeyInMetadata {
+		keyname = ac.Name()
+	}
+
+	mach := &machine{
+		cluster: ac,
+		mach:    device,
+	}
+
+	mach.dir = filepath.Join(ac.RuntimeConf().OutputDir, mach.ID())
+	if err := os.Mkdir(mach.dir, 0777); err != nil {
+		mach.Destroy()
+		return nil, err
+	}
+
+	confPath := filepath.Join(mach.dir, "user-data")
+	if err := conf.WriteFile(confPath); err != nil {
+		mach.Destroy()
+		return nil, err
+	}
+
+	if mach.journal, err = platform.NewJournal(mach.dir); err != nil {
+		mach.Destroy()
+		return nil, err
+	}
+
+	if err := mach.journal.Start(context.TODO(), mach); err != nil {
+		mach.Destroy()
+		return nil, err
+	}
+
+	if err := platform.CheckMachine(mach); err != nil {
+		mach.Destroy()
+		return nil, fmt.Errorf("machine %q failed basic checks: %v", mach.ID(), err)
+	}
+
+	if err := platform.EnableSelinux(mach); err != nil {
+		mach.Destroy()
+		return nil, err
+	}
+
+	ac.AddMach(mach)
+
+	return mach, nil
+}
+
+func (ac *cluster) Destroy() error {
+	if !ac.RuntimeConf().NoSSHKeyInMetadata {
+		if err := ac.api.DeleteKey(ac.Name()); err != nil {
+			return err
+		}
+	}
+
+	return ac.BaseCluster.Destroy()
+}
+
+func (ac *cluster) vmname() string {
+	b := make([]byte, 5)
+	rand.Read(b)
+	return fmt.Sprintf("%s-%x", ac.Name()[0:13], b)
```

Comment:
If ESX doesn't have length limits on the VM name, you should probably drop the `[0:13]`.